deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/parallel/ParallelExistingMiniBatchDataSetIterator.java

```diff
@@ -0,0 +1,397 @@
+     */
+    public ParallelExistingMiniBatchDataSetIterator(File rootDir) {
+        this(rootDir, DEFAULT_PATTERN);
+    }
+
+    public ParallelExistingMiniBatchDataSetIterator(File rootDir, int numThreads) {
+        this(rootDir, DEFAULT_PATTERN, numThreads);
+    }
+
+    /**
+     *
+     * @param rootDir    The root directory to use
+     * @param pattern    The filename pattern to use. Used with {@code String.format(pattern,idx)}, where idx is an
+     *                   integer, starting at 0.
+     */
+    public ParallelExistingMiniBatchDataSetIterator(File rootDir, String pattern) {
+        this(rootDir, pattern, 2);
+    }
+
+    public ParallelExistingMiniBatchDataSetIterator(File rootDir, String pattern, int numThreads) {
+        this(rootDir, pattern, numThreads, 8);
+    }
+
+    public ParallelExistingMiniBatchDataSetIterator(File rootDir, String pattern, int numThreads, int bufferSize) {
+        this(rootDir, pattern, numThreads, bufferSize, true);
+    }
+
+    public ParallelExistingMiniBatchDataSetIterator(File rootDir, String pattern, int numThreads, int bufferSize, boolean useWorkspaces) {
+        if (numThreads < 2)
+            numThreads = 2;
+
+        this.numThreads = numThreads;
+        this.rootDir = rootDir;
+        totalBatches = rootDir.list().length;
+        this.pattern = pattern;
+        this.bufferSize = bufferSize;
+        this.useWorkspaces = useWorkspaces;
+
+        this.buffer = new LinkedBlockingQueue<>(this.bufferSize);
+
+        this.executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(numThreads, new ThreadFactory() {
+            @Override
+            public Thread newThread(Runnable r) {
+                Thread t = Executors.defaultThreadFactory().newThread(r);
+                // PEMBDSI. I like it.
+                t.setName("PEMBDSI pool thread ");
+                t.setDaemon(true);
+                return t;
+            }
+        });
+
+        this.thread = new AsyncDispatcherThread();
+        Nd4j.getAffinityManager().attachThreadToDevice(thread, Nd4j.getAffinityManager().getDeviceForCurrentThread());
+
+        this.thread.start();
+    }
+
+    @Override
+    public DataSet next(int num) {
+        throw new UnsupportedOperationException("Unable to load custom number of examples");
+    }
+
+    @Override
+    public int totalExamples() {
+        throw new UnsupportedOperationException();
+
+    }
+
+    @Override
+    public int inputColumns() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int totalOutcomes() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean resetSupported() {
+        return true;
+    }
+
+    @Override
+    public boolean asyncSupported() {
+        return true;
+    }
+
+    @Override
+    public void reset() {
+        shouldWork.set(false);
+        thread.interrupt();
+        thread.shutdown();
+        buffer.clear();
```

Comment:
no, thread.shutdown() call is blocking. so buffer.clear() is guaranteed to happen only after current thread is dead.